TypeError in to_pydict with optional fields

See https://github.com/danielgtaylor/python-betterproto/issues/475#issuecomment-1512296041

<details>
<summary>Outdated</summary>

I am using betterproto to load a ton of messages at once from a binary file, however this operation is very slow so I wanted to cache/parse it into json using `json.dump` with `to_dict`, and then load it directly from json. Problem is that `from_json` is also very slow, so my idea was to use `to_dict(casing=Casing.SNAKE, include_default_values=true)` to generate a dictionary that matches the attributes of the message dataclass, and then wrap it in an utility class that redirects dict.attribute calls to dict["attribute"] so that for the consumer (which is using the message dataclasses exclusively for their data) there's no difference.

However the problem is that when serializing to dictionary, enum values are serialized based on their name instead of value. Maybe an optional parameter can be added to change this behaviour? Thank you


<details>
<summary>Utility dictionary class for redirecting attributes to items</summary>

```py
class AttrDict(dict):
    def __getattr__(self, key):
        try:
            return self[key]
        except KeyError:
            raise AttributeError(f"'{self.__class__.__name__}' object has no attribute '{key}'")

    def __setattr__(self, key, value):
        self[key] = value

    def __getitem__(self, key):
        value = super().__getitem__(key)
        if isinstance(value, dict) and not isinstance(value, AttrDict):
            value = AttrDict(value)
            self[key] = value
        if isinstance(value, list) and any(isinstance(x, dict) for x in value):
            value = [AttrDict(x) if isinstance(x, dict) else x for x in value]
            self[key] = value
        return value
```

</details>

</details>

I was originally trying with 3.10 on Mac OS, now just tried with Python 3.11.3, Poetry 1.4.2 on Ubuntu 22.04:

```
  third_party/protobuf/src/google/protobuf/util/internal/datapiece.cc:57:36: warning: comparison of integer expressions of different signedness: ‘long int’ and ‘long unsigned int’ [-Wsign-compare]
     57 |       MathUtil::Sign<From>(before) == MathUtil::Sign<To>(after)) {
        |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
  gcc -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DHAVE_PTHREAD=1 -I. -Igrpc_root -Igrpc_root/include -Ithird_party/protobuf/src -I/tmp/tmptrqf78k0/.venv/include -I/root/.asdf/installs/python/3.11.3/include/python3.11 -c third_party/protobuf/src/google/protobuf/util/message_differencer.cc -o build/temp.linux-x86_64-cpython-311/third_party/protobuf/src/google/protobuf/util/message_differencer.o -std=c++14 -fno-wrapv -frtti
  gcc -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DHAVE_PTHREAD=1 -I. -Igrpc_root -Igrpc_root/include -Ithird_party/protobuf/src -I/tmp/tmptrqf78k0/.venv/include -I/root/.asdf/installs/python/3.11.3/include/python3.11 -c third_party/protobuf/src/google/protobuf/util/internal/protostream_objectwriter.cc -o build/temp.linux-x86_64-cpython-311/third_party/protobuf/src/google/protobuf/util/internal/protostream_objectwriter.o -std=c++14 -fno-wrapv -frtti
  third_party/protobuf/src/google/protobuf/util/internal/utility.cc: In function ‘void google::protobuf::util::converter::InitWellKnownTypes()’:
  third_party/protobuf/src/google/protobuf/util/internal/utility.cc:348:21: warning: comparison of integer expressions of different signedness: ‘int’ and ‘long unsigned int’ [-Wsign-compare]
    348 |   for (int i = 0; i < GOOGLE_ARRAYSIZE(well_known_types_name_array_); ++i) {
  third_party/protobuf/src/google/protobuf/util/internal/protostream_objectwriter.cc: In member function ‘void google::protobuf::util::converter::ProtoStreamObjectWriter::AnyWriter::StartAny(const google::protobuf::util::converter::DataPiece&)’:
  third_party/protobuf/src/google/protobuf/util/internal/protostream_objectwriter.cc:380:21: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<google::protobuf::util::converter::ProtoStreamObjectWriter::AnyWriter::Event>::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
    380 |   for (int i = 0; i < uninterpreted_events_.size(); ++i) {
        |                   ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  gcc -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DHAVE_PTHREAD=1 -I. -Igrpc_root -Igrpc_root/include -Ithird_party/protobuf/src -I/tmp/tmptrqf78k0/.venv/include -I/root/.asdf/installs/python/3.11.3/include/python3.11 -c third_party/protobuf/src/google/protobuf/compiler/js/well_known_types_embed.cc -o build/temp.linux-x86_64-cpython-311/third_party/protobuf/src/google/protobuf/compiler/js/well_known_types_embed.o -std=c++14 -fno-wrapv -frtti
  gcc -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DHAVE_PTHREAD=1 -I. -Igrpc_root -Igrpc_root/include -Ithird_party/protobuf/src -I/tmp/tmptrqf78k0/.venv/include -I/root/.asdf/installs/python/3.11.3/include/python3.11 -c third_party/protobuf/src/google/protobuf/timestamp.pb.cc -o build/temp.linux-x86_64-cpython-311/third_party/protobuf/src/google/protobuf/timestamp.pb.o -std=c++14 -fno-wrapv -frtti
  gcc -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DHAVE_PTHREAD=1 -I. -Igrpc_root -Igrpc_root/include -Ithird_party/protobuf/src -I/tmp/tmptrqf78k0/.venv/include -I/root/.asdf/installs/python/3.11.3/include/python3.11 -c third_party/protobuf/src/google/protobuf/wire_format_lite.cc -o build/temp.linux-x86_64-cpython-311/third_party/protobuf/src/google/protobuf/wire_format_lite.o -std=c++14 -fno-wrapv -frtti
  gcc -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DHAVE_PTHREAD=1 -I. -Igrpc_root -Igrpc_root/include -Ithird_party/protobuf/src -I/tmp/tmptrqf78k0/.venv/include -I/root/.asdf/installs/python/3.11.3/include/python3.11 -c third_party/protobuf/src/google/protobuf/descriptor.pb.cc -o build/temp.linux-x86_64-cpython-311/third_party/protobuf/src/google/protobuf/descriptor.pb.o -std=c++14 -fno-wrapv -frtti
  error: command '/usr/bin/gcc' failed with exit code 1
  

  at ~/.local/pipx/venvs/poetry/lib/python3.11/site-packages/poetry/installation/chef.py:152 in _prepare
      148│ 
      149│                 error = ChefBuildError("\n\n".join(message_parts))
      150│ 
      151│             if error is not None:
    → 152│                 raise error from None
      153│ 
      154│             return path
      155│ 
      156│     def _prepare_sdist(self, archive: Path, destination: Path | None = None) -> Path:

Note: This error originates from the build backend, and is likely not a problem with poetry but with grpcio-tools (1.48.2) not supporting PEP 517 builds. You can verify this by running 'pip wheel --use-pep517 "grpcio-tools (==1.48.2)"'.

```

There is a lot more output before that as well, let me know if it's helpful.

--- a/src/betterproto/__init__.py
+++ b/src/betterproto/__init__.py
@@ -1373,6 +1373,9 @@ class Message(ABC):
                     value = [i.to_pydict(casing, include_default_values) for i in value]
                     if value or include_default_values:
                         output[cased_name] = value
+                elif value is None:
+                    if include_default_values:
+                        output[cased_name] = None
                 elif (
                     value._serialized_on_wire
                     or include_default_values

--- a/tests/test_features.py
+++ b/tests/test_features.py
@@ -265,6 +265,32 @@ def test_optional_flag():
     assert Request().parse(b"\n\x00").flag is False
 
 
+def test_optional_datetime_to_dict():
+    @dataclass
+    class Request(betterproto.Message):
+        date: Optional[datetime] = betterproto.message_field(1, optional=True)
+
+    # Check dict serialization
+    assert Request().to_dict() == {}
+    assert Request().to_dict(include_default_values=True) == {"date": None}
+    assert Request(date=datetime(2020, 1, 1)).to_dict() == {
+        "date": "2020-01-01T00:00:00Z"
+    }
+    assert Request(date=datetime(2020, 1, 1)).to_dict(include_default_values=True) == {
+        "date": "2020-01-01T00:00:00Z"
+    }
+
+    # Check pydict serialization
+    assert Request().to_pydict() == {}
+    assert Request().to_pydict(include_default_values=True) == {"date": None}
+    assert Request(date=datetime(2020, 1, 1)).to_pydict() == {
+        "date": datetime(2020, 1, 1)
+    }
+    assert Request(date=datetime(2020, 1, 1)).to_pydict(
+        include_default_values=True
+    ) == {"date": datetime(2020, 1, 1)}
+
+
 def test_to_json_default_values():
     @dataclass
     class TestMessage(betterproto.Message):

